### ISSUE TITLE
Grader caches old exercise templates and can not update them without restarting the whole server

### ISSUE THREAD

**Issue by Mankro** [apluslms/mooc-grader#105]
This has been an obscure problem in Tampere. The MOOC-Grader server seems to be unable to load new versions of **exercise templates** when the course is updated via git. It looks like the HTML file is updated in the server file system, but it does not update in the MOOC-Grader code. Then, the server shows an outdated version of the exercise to the students even though the teacher has updated it in the course git repo.

The exercise (HTML) template refers to the `template` field in the exercise config.yaml.

There is also an open question if this caching problem affects other fields in the exercise config.yaml. For example, can the same problem occur with the `instructions` field when the default HTML template is used?

Some code that is related to the caching in the MOOC-Grader:

* https://github.com/apluslms/mooc-grader/blob/6e5bc6b177debefcb404bbf4b0c2f77715e9cf2a/access/types/stdasync.py#L49
* https://github.com/apluslms/mooc-grader/blob/6e5bc6b177debefcb404bbf4b0c2f77715e9cf2a/access/views.py#L293
* https://github.com/apluslms/mooc-grader/blob/6e5bc6b177debefcb404bbf4b0c2f77715e9cf2a/access/config.py#L181

**Comment by Mankro** [comment]
@jykesaviaGNH and @PasiSa this is what we talked about today.

**Comment by Mankro** [comment]
I have now found another case that could be caused by the same problem in the MOOC-Grader. The course uses the index.yaml format (no RST files) and the default HTML template of the `view_type: access.types.stdsync.createForm`. The exercise description is written in the exercise yaml file in the `instructions` field. Sometimes, the server seems to be unable to load updates to the instructions and it keeps showing the old version of the instructions.

The course also uses base yaml config files, that is to say, the exercise yaml file uses the `ìnclude`, `file` and `template_context` fields.

**Comment by Mankro** [comment]
I fixed a problem with included base configuration templates (the `include` field in the exercise config.yaml, which is usually not used at all). MOOC-Grader reloads course configurations to the cache if the YAML file has a newer modification timestamp than previously. However, it didn't check the timestamp of the included base config template. This has been fixed.

From the commit message https://github.com/apluslms/mooc-grader/commit/2fcd5ad9e74ac97a4563b9af3274b462002a14de:
> The course configuration cache checks if the configurations of
the requested exercise have changed in the file system by
comparing the modification timestamp of the exercise configuration
file (often named config.yaml) to the timestamp stored in the cache.
However, some settings may be changed without modifying
the config.yaml file itself. This commit detects the changes in
the included base configuration template by checking its file
modification timestamp too. The base configuration template
is defined with the "include" field (if it is used).

There were suspicions that there are issues in reloading changes in exercise HTML templates (the `template` field in config.yaml). However, I could not detect any bug with those. MOOC-Grader does not cache HTML templates and it uses the version of the template that exists in the file system. I did notice that the A+ frontend caches exercise contents and A+ would typically serve outdated exercise content from the cache if the exercise template is changed in the grader without importing the course into A+ again. The teacher may clear the content cache in A+ by pressing the "clear content cache" button in the A+ Edit course page.

In the future, MOOC-Grader could hopefully send update messages to A+ so that the teacher would not need to manually clear the cache in A+ after changing exercise templates (or the exercise instructions in config.yaml).
